File: ident-uuid/src/main/scala/net/nicktelford/ident/uuid/rfc4122/RandomUUID.scala
package net.nicktelford.ident.uuid.rfc4122

import net.nicktelford.ident.uuid._

/** Companion object for generating RandomUUIDs. */
object RandomUUID extends RFC4122Factory[RandomUUID] {

  /** @see UUIDFactory.MaxValue */
  val MaxValue: RandomUUID = new RandomUUID(encodeMSB(-1L), encodeLSB(-1, -1))

  /** @see UUIDFactory.MinValue */
  val MinValue: RandomUUID = new RandomUUID(encodeMSB(0), encodeLSB(0, 0))

  /** @see RFC4122Factory.version */
  val version: Short = 4

  /** Generates a RandomUUID */
  def apply(): RandomUUID = new RandomUUID(randomMSB, randomLSB)
}

/** A Version 4, randomly generated UUID. */
case class RandomUUID(msb: Long, lsb: Long) extends UUID with RFC4122 {

  /** @see RFC4122.version */
  val version: Short = 4
}

File: ident-java/src/main/java/net/nicktelford/ident/rfc4122/RandomUUID.java
package net.nicktelford.ident.rfc4122;

/**
 * An RFC4122 Version 4, randomly generated UUID.
 *
 * @see <a href="http://www.ietf.org/rfc/rfc4122">RFC4122</a>
 */
public class RandomUUID extends RFC4122UUID {

    public static final short VERSION = 4;
    private static final Factory FACTORY = new Factory(VERSION);

    public static final RandomUUID MAX_VALUE
            = new RandomUUID(FACTORY.encodeMSB(-1L), FACTORY.encodeLSB(-1));
    public static final RandomUUID MIN_VALUE
            = new RandomUUID(FACTORY.encodeMSB(0), FACTORY.encodeLSB(0));

    public static RandomUUID generate() {
        return new RandomUUID(FACTORY.randomMSB(), FACTORY.randomLSB());
    }

    RandomUUID(final long msb, final long lsb) {
        super(msb, lsb);
    }

    @Override
    public short version() {
        return VERSION;
    }
}
